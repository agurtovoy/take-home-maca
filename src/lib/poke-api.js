import useSWRImmutable from "swr/immutable";
import useSWRInfinite from "swr/infinite";
import { request } from "graphql-request";
export { gql } from "graphql-request";

const fetcher = ([query, args]) => {
    console.log("fetching query", query, { args });
    return request("https://beta.pokeapi.co/graphql/v1beta", query, args);
};

export const usePokeAPI = (url, options) => useSWRImmutable(url, fetcher, options);
export const usePokeAPIInfinite = (getKey, options) =>
    useSWRInfinite(getKey, fetcher, {
        revalidateFirstPage: false,
        revalidateOnFocus: false,
        revalidateIfStale: false,
        revalidateOnReconnect: false,
        keepPreviousData: true,
        ...options,
    });

export default usePokeAPI;
