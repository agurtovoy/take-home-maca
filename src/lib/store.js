import { createStore } from "zustand";

export const createFiltersStore = ({ typeFiltersList, logging }) => {
    const typeFilters = (typeFiltersList || []).reduce((result, { id, name }) => ({ ...result, [id]: name }), {});
    const storeDef = (set, get) => ({
        typeFilters: { ...typeFilters },
        typeSelection: {},
        selectedTypes: () => Object.entries(get().typeSelection).filter(s => s[1]),
        selectedTypeIds: () =>
            get()
                .selectedTypes()
                .map(([id]) => parseInt(id, 10)),
        selectedTypeNames: () =>
            get()
                .selectedTypeIds()
                .map(id => typeFilters[id]),
        selectType: (id, selected) =>
            set(({ typeSelection }) => ({ typeSelection: { ...typeSelection, [id]: selected } })),
        clearAll: () => set({ typeSelection: {} }),
    });

    return logging ? createStore(withLogging(storeDef)) : createStore(storeDef);
};

const withLogging = config => (set, get, api) =>
    config(
        args => {
            console.log("-> applying:", args);
            set(args);
            console.log("-> new state:", get());
        },
        get,
        api
    );
