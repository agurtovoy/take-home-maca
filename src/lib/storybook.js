export const configure = (Story, args, params) => {
    const result = Story.bind({});
    result.args = args;
    result.parameters = params;
    return result;
};

// roughly based on https://dev.to/admitkard/storybook-controls-the-easy-way-5g4p
export const select = (options, defaultValue, { type } = {}) => ({
    options,
    defaultValue: defaultValue || options[0],
    control: { type: type || "select" },
});

export const boolean = (defaultValue, { type } = {}) => ({
    defaultValue: defaultValue || false,
    control: { type: type || "boolean" },
});

export const text = (defaultValue, { type } = {}) => ({
    defaultValue: defaultValue || "",
    control: { type: type || "text" },
});

export const number = (defaultValue, { type, min, max, step } = {}) => ({
    defaultValue: defaultValue || 0,
    control: { type: type || "number", min: min || -Infinity, max: max || Infinity, step: step || 1 },
});

export const date = (defaultValue, { type } = {}) => ({
    defaultValue: defaultValue || new Date(),
    control: { type: type || "date" },
});

export const object = (defaultValue, { type, min, max, step } = {}) => ({
    defaultValue: defaultValue || {},
    control: { type: type || "object" },
});

export const action = actionName => ({
    action: actionName,
});
