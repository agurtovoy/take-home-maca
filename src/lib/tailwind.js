import { useMedia } from "react-use";
// see https://twitter.com/silvenon/status/1375149895645036549
import * as screens from "@/styles/screens.module.css";

const remapKey = { "2xl": "xxl" };

export const useScreen = screen => {
    screen = remapKey[screen] || screen;
    return useMedia(`(min-width: ${screens[screen]})`);
};
