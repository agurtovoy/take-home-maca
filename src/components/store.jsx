import { useRef } from "react";
import { createContext, useContext } from "react";
import { useStore as useStore_ } from "zustand";

import { createFiltersStore } from "@/lib/store";

const StoreContext = createContext(null);

export const StoreProvider = ({ children, ...props }) => {
    const storeRef = useRef();
    if (!storeRef.current) {
        storeRef.current = createFiltersStore(props);
    }
    return <StoreContext.Provider value={storeRef.current}>{children}</StoreContext.Provider>;
};

export const useStore = (selector, equalityFn) => {
    const store = useContext(StoreContext);
    if (!store) throw new Error("Missing StoreContext.Provider in the tree");
    return useStore_(store, selector, equalityFn);
};
