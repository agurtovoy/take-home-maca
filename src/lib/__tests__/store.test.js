import { expect, describe, test } from "vitest";
import { createFiltersStore } from "@/lib/store";

const filters = [
    {
        name: "normal",
        id: 101,
    },
    {
        name: "fighting",
        id: 17,
    },
    {
        name: "flying",
        id: 42,
    },
];

describe("correctly initialized", () => {
    test("list of types", () => {
        const store = createFiltersStore({
            typeFiltersList: filters,
        }).getState();

        expect(store.typeFilters).toMatchSnapshot();
        expect(store.typeSelection).toStrictEqual({});
    });

    test("null/undefined", () => {
        expect(createFiltersStore({ typeFiltersList: null }).getState().typeFilters).toStrictEqual({});
        expect(createFiltersStore({}).getState().typeFilters).toStrictEqual({});
    });
});

describe("selection", () => {
    test("set/unset", () => {
        const store = createFiltersStore({
            typeFiltersList: filters,
        }).getState();

        expect(store.selectedTypeIds()).toStrictEqual([]);
        expect(store.selectedTypeNames()).toStrictEqual([]);

        store.selectType(17, true);
        store.selectType(42, false);
        store.selectType(101, true);
        expect(store.selectedTypeIds()).toMatchSnapshot();
        expect(store.selectedTypeNames()).toMatchSnapshot();

        store.clearAll();
        expect(store.selectedTypeIds()).toStrictEqual([]);
        expect(store.selectedTypeNames()).toStrictEqual([]);
    });
});
