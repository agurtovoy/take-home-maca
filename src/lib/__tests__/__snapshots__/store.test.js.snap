// Vitest Snapshot v1

exports[`correctly initialized > list of types 1`] = `
{
  "101": "normal",
  "17": "fighting",
  "42": "flying",
}
`;

exports[`selection > set/unset 1`] = `
[
  17,
  101,
]
`;

exports[`selection > set/unset 2`] = `
[
  "fighting",
  "normal",
]
`;
